File: Python/medium/49_group_anagrams.py
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        values = {}
        for str in strs:
            chars = [x for x in str]
            chars.sort()
            key = "".join(chars)
            if key in values:
                values[key].append(str)
            else:
                values[key] = [str]

        return list(values.values())


s = Solution()
arr = s.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"])
print(arr)

File: C++/medium/49_group_anagrams.cc
#include <iostream>

using namespace std;

class Solution
{
public:
    vector<vector<string>> groupAnagrams(vector<string> &strs)
    {
        vector<vector<string>> res;
        unordered_map<string, vector<string>> m;

        for (string s : strs)
        {
            string t = s;
            sort(t.begin(), t.end());
            m[t].push_back(s);
        }

        for (auto a : m)
        {
            res.push_back(a.second);
        }

        return res;
    }
};
